linera-execution/src/wasm/async_boundary.rs

```diff
@@ -5,156 +5,240 @@
-    /// to hang.
-    ///
-    /// # Panics
-    ///
-    /// If the `context` does not contain a valid exclusive task [`Waker`] reference, or if this
-    /// future is polled concurrently in different tasks.
-    pub fn poll(&self, waker: &mut WakerForwarder) -> Poll<Output> {
-        let waker_reference = waker
-            .0
-            .try_lock()
-            .expect("Unexpected concurrent application call");
-
-        let mut context = Context::from_waker(
-            waker_reference
-                .as_ref()
-                .expect("Application called without an async task context"),
-        );
-
-        let mut future = self
-            .future
-            .try_lock()
-            .expect("Application can't call the future concurrently because it's single threaded");
-
-        future.as_mut().poll(&mut context)
+    /// Executes the `future`, polling it as requested by the [`PollSender`] until it completes.
+    fn run_poll_loop(mut self, future: Future) {
+        while let Ok(response_sender) = self.poll_requests.recv() {
+            let response = future.poll(&self.context.application, &mut self.context.store);
+            let finished = response.is_ready();
+
+            let _ = response_sender.send(PollResponse(response));
+
+            if finished {
+                break;
+            }
+        }
+    }
+
+    /// Waits for the first poll request sent by the [`PollSender`], and immediately responds with
+    /// the error obtained when attempting to create the `Future` instance.
+    fn notify_creation_error(self, error: ExecutionError) {
+        if let Ok(response_sender) = self.poll_requests.recv() {
+            let _ = response_sender.send(PollResponse(Poll::Ready(Err(error))));
+        }
     }
 }
 
-/// A future implemented in a Wasm module.
-pub enum GuestFuture<'context, Future, Application>
-where
-    Application: ApplicationRuntimeContext,
-{
-    /// The Wasm module failed to create an instance of the future.
-    ///
-    /// The error will be returned when this [`GuestFuture`] is polled.
-    FailedToCreate(Option<Application::Error>),
+/// A wrapper type representing the response sent from a future actor to a poll request sent from a
+/// [`PollSender`].
+///
+/// This helps to simplify some types used, avoiding some Clippy lints.
+struct PollResponse<Output>(Poll<Result<Output, ExecutionError>>);
+
+/// An abstraction over a [`Future`] running as an actor on a non-asynchronous thread.
+///
+/// This type implements [`Future`] and sends poll requests to the actor implementation. When the
+/// actor finishes, it sends back the result to this type, which then returns it.
+///
+/// Poll requests may not be sent to the implementation if it would cause non-deterministic
+/// execution (as controlled by the [`HostFutureQueue`]).
+pub struct PollSender<Output> {
+    host_future_queue: Option<HostFutureQueue>,
+    poll_requester: std::sync::mpsc::Sender<oneshot::Sender<PollResponse<Output>>>,
```

Comment:
nit: We may want to stick to the basic name convention `sender: mpsc::Sender<..>` or `poll_request_sender: mpsc::Sender<..>`. There could also be a case for defining `struct PollRequest { response: oneshot::Sender<PollResponse<Output>> }` out of consistency.

Reply:
Done.